third_party/move/move-compiler-v2/src/pipeline/explicit_drop.rs

```diff
@@ -0,0 +1,197 @@
+use super::{
+    livevar_analysis_processor::{LiveVarAnnotation, LiveVarInfoAtCodeOffset},
+    reference_safety_processor::{LifetimeAnnotation, LifetimeInfoAtCodeOffset},
+};
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv, ty::Type};
+use move_stackless_bytecode::{
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AssignKind, AttrId, Bytecode, Operation},
+};
+use std::collections::BTreeSet;
+
+pub struct ExplicitDrop {}
+
+impl FunctionTargetProcessor for ExplicitDrop {
+    fn process(
+        &self,
+        _targets: &mut FunctionTargetsHolder,
+        fun_env: &FunctionEnv,
+        mut data: FunctionData,
+        _scc_opt: Option<&[FunctionEnv]>,
+    ) -> FunctionData {
+        if fun_env.is_native() {
+            return data;
+        }
+        let target = FunctionTarget::new(fun_env, &data);
+        let mut transformer = ExplicitDropTransformer::new(target);
+        transformer.transform();
+        data.code = transformer.transformed;
+        data.annotations.remove::<LiveVarAnnotation>();
+        data.annotations.remove::<LifetimeAnnotation>();
+        data
+    }
+
+    fn name(&self) -> String {
+        "ExplicitDrop".to_owned()
+    }
+}
+
+struct ExplicitDropTransformer<'a> {
+    target: FunctionTarget<'a>,
+    // result of the transformation
+    transformed: Vec<Bytecode>,
+    live_var_annot: &'a LiveVarAnnotation,
+    lifetime_annot: &'a LifetimeAnnotation,
+}
+
+impl<'a> ExplicitDropTransformer<'a> {
+    pub fn new(target: FunctionTarget<'a>) -> Self {
+        let live_var_annot = target
+            .get_annotations()
+            .get::<LiveVarAnnotation>()
+            .expect("livevar annotation");
+        let lifetime_annot = target
+            .get_annotations()
+            .get::<LifetimeAnnotation>()
+            .expect("lifetime annotation");
+        ExplicitDropTransformer {
+            target,
+            transformed: Vec::new(),
+            live_var_annot,
+            lifetime_annot,
+        }
+    }
+
+    /// Add explicit drop instructions
+    /// note that this will invalidate existing analyses
+    pub fn transform(&mut self) {
+        self.drop_unused_args();
+        for (code_offset, bytecode) in self.target.get_bytecode().to_vec().iter().enumerate() {
+            self.emit_bytecode(bytecode.clone());
+            self.explicate_drops_at(code_offset as CodeOffset, bytecode);
+        }
+    }
+
+    // add drops at given code offset
+    fn explicate_drops_at(&mut self, code_offset: CodeOffset, bytecode: &Bytecode) {
```

Comment:
"explicate" is a pretty unusual English word, and this name is confusing.  Perhaps `add_drops_at` is what you mean?  Or `insert_explicit_drops_at`?

Reply:
I've changed it to explicit in the latest commit. Btw, I used explicate because a compiler book I used uses "explicate control" for a pass which makes execution order explicit.

Reply:
Interesting.  I think this may have been first used [in 2005 ](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=bf91e6b4a257faf9a028cef32d2e00364a846ea4) as a verb by a Dutch author, then popularized as the name of a compiler pass in a [book from U. of Indiana](https://brinckerhoff.org/clements/2194-csc431/essentials-of-compilation.pdf).  It makes some sense to use "explicate control flow" as a more concise way of saying "make control flow explicit" but I think naming the pass "explicate control" wasn't a great idea, as it seems to be trying to use "explicate" as an adjective instead of a verb.  Compare with "live variable analysis" or "register allocation" or "reaching definitions".  "Control explication" would have made more sense as a compiler pass.